physical/raft/transaction.go

```diff
@@ -0,0 +1,605 @@
+	return ret
+}
+
+type raftTxnUpdateRecord struct {
+	// If this record exists but Contents is nil, the entry was deleted.
+	Contents *physical.Entry
+}
+
+type RaftTransaction struct {
+	b              *RaftBackend
+	l              sync.RWMutex
+	tx             *bbolt.Tx
+	updates        map[string]*raftTxnUpdateRecord
+	log            *LogData
+	writable       bool
+	haveWritten    bool
+	haveFinishedTx bool
+}
+
+var _ physical.Transaction = &RaftTransaction{}
+
+func (b *RaftBackend) newTransaction(ctx context.Context, writable bool) (*RaftTransaction, error) {
+	// Grab a transaction permit pool entry so that we can limit the number of
+	// concurrent transactions. Also grab a read lock in the underlying FSM
+	// to prevent key changes from occurring while a transaction is ongoing.
+	// These will be released when we finish this transaction.
+	b.txnPermitPool.Acquire()
+	b.fsm.l.RLock()
+
+	// All transactions are read-only; this gives us a consistent view of
+	// storage but means we need to track writes ourselves.
+	tx, err := b.fsm.db.Begin(false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start underlying bbolt transaction: %w", err)
+	}
+
+	return &RaftTransaction{
+		b:       b,
+		tx:      tx,
+		updates: make(map[string]*raftTxnUpdateRecord),
+		log: &LogData{
+			Operations: []*LogOperation{
+				{
+					OpType: beginTxOp,
+				},
+			},
+		},
+		writable: writable,
+	}, nil
+}
+
+func (t *RaftTransaction) Put(ctx context.Context, entry *physical.Entry) error {
+	if !t.writable {
+		return physical.ErrTransactionReadOnly
+	}
+
+	t.l.Lock()
+	defer t.l.Unlock()
+	t.haveWritten = true
+
+	// If we haven't modified this entry within the scope of this
+	// transaction, read the value of this entry so we can hash it.
+	if _, present := t.updates[entry.Key]; !present {
+		// It is safe to go to the underlying transaction here as we
+		// hold an exclusive write lock here and so there's no parallel
+		// writers to the same key.
+		value := t.tx.Bucket(dataBucketName).Get([]byte(entry.Key))
+		contentsHash, err := createVerificationEntry(entry.Key, value)
+		if err != nil {
+			return err
+		}
+
+		// Verify the entry prior to updating it, when it comes time for raft
+		// application.
+		t.log.Operations = append(t.log.Operations, &LogOperation{
+			OpType: verifyReadOp,
+			Key:    entry.Key,
+			Value:  contentsHash,
+		})
+	}
+
+	update := &raftTxnUpdateRecord{
+		// Caller may mutate their entry after we accept it, so create a new
+		// one for the cache.
+		Contents: &physical.Entry{
+			Key:   entry.Key,
+			Value: cloneBytes(entry.Value),
+		},
+	}
+
+	// Do the update in the transaction, adding it to the future raft log.
+	t.updates[entry.Key] = update
+	t.log.Operations = append(t.log.Operations, &LogOperation{
+		OpType: putOp,
+		Key:    entry.Key,
+		Value:  entry.Value,
+	})
+
+	return nil
+}
+
+func (t *RaftTransaction) Get(ctx context.Context, key string) (*physical.Entry, error) {
+	t.l.RLock()
```

Comment:
This should probably be dropped to a regular mutex, as even for list/read operations, we append to `t.log.Operations`, so we can't necessarily do those in parallel.